### PR TITLE
5811 - Serve Petition from Message Detail

### PR DESCRIPTION
**Pull request by matthopson**
Screen Shot 2020-07-17 at 12 08 01 PM

Screen Shot 2020-07-17 at 12 08 22 PM

### CODE REVIEW

File: web-client/src/presenter/sequences/closeModalAndReturnToDocumentQCSequence.js [hunk 1]
```diff
@@ -1,7 +1,13 @@
 import { clearModalAction } from '../actions/clearModalAction';
+import { followRedirectAction } from '../actions/followRedirectAction';
 import { navigateToDocumentQCAction } from '../actions/navigateToDocumentQCAction';
 
+// TODO: Maybe rename this since it's not _always_ returning to DocumentQC if redirect is set
```
Comment (matthopson) [inline web-client/src/presenter/sequences/closeModalAndReturnToDocumentQCSequence.js]: Anyone have any feels on this?

File: web-client/src/presenter/computeds/messageDocumentHelper.js [hunk 2]
```diff
@@ -82,6 +87,9 @@ export const messageDocumentHelper = (get, applicationContext) => {
   const showServeCourtIssuedDocumentButton =
     showNotServed && isCourtIssuedDocument && permissions.SERVE_DOCUMENT;
 
+  const showServePetitionButton =
+    showNotServed && isPetitionDocument && permissions.SERVE_DOCUMENT;
```
Comment (rachaelparris) [inline web-client/src/presenter/computeds/messageDocumentHelper.js]: Should be `permissions.SERVE_PETITION`

File: web-client/src/presenter/computeds/messageDocumentHelper.js [hunk 3]
```diff
@@ -99,5 +107,6 @@ export const messageDocumentHelper = (get, applicationContext) => {
       showApplyEditSignatureButtonForRole && showEditSignatureButtonForDocument,
     showNotServed,
     showServeCourtIssuedDocumentButton,
+    showServePetitionButton: showServePetitionButton,
```
Comment (rachaelparris) [inline web-client/src/presenter/computeds/messageDocumentHelper.js]: Nit: can be shortened to just `showServePetitionButton`

File: web-client/src/presenter/computeds/messageDocumentHelper.test.js [hunk 4]
```diff
@@ -820,4 +820,94 @@ describe('messageDocumentHelper', () => {
       expect(result.showServeCourtIssuedDocumentButton).toEqual(true);
     });
   });
+
+  describe('showServePetitionButton', () => {
+    it('should be false is the document is a Petition document and the user has SERVE_PETITION permission', () => {
```
Comment (rachaelparris) [inline web-client/src/presenter/computeds/messageDocumentHelper.test.js]: Nit: `should be false *if*` (same on other tests)

Comment (rachaelparris) [reply web-client/src/presenter/computeds/messageDocumentHelper.test.js]: Also, maybe add "and the petition is served" for clarity?

File: web-client/src/presenter/computeds/messageDocumentHelper.js (hunk 2 above)
Comment (matthopson) [reply web-client/src/presenter/computeds/messageDocumentHelper.js]: d4fb01f72b8e246ff0f2a5ac2276d09f
After I went through all that effort... lol

File: web-client/src/presenter/sequences/closeModalAndReturnToDocumentQCSequence.js (hunk 1 above)
Comment (rachaelparris) [reply web-client/src/presenter/sequences/closeModalAndReturnToDocumentQCSequence.js]: How about `closeModalAndNavigateSequence`?

File: web-client/src/presenter/sequences/navigateBackSequence.js [hunk 5]
```diff
@@ -7,9 +8,15 @@ import { stopShowValidationAction } from '../actions/stopShowValidationAction';
 export const navigateBackSequence = [
   clearAlertsAction,
   stopShowValidationAction,
-  getHasAlternateBackLocationAction,
+  followRedirectAction,
   {
-    false: [navigateBackAction],
-    true: [navigateToPathAction],
+    default: [
+      getHasAlternateBackLocationAction, // TODO: For refactor - could probably use the redirect implementation instead (see noticeGenerationCompleteSequence)
```
Comment (rachaelparris) [inline web-client/src/presenter/sequences/navigateBackSequence.js:14]: Could you add this to the refactoring task list in the story or in Trello? I don't want this TODO to get missed.

File: web-client/src/views/Messages/MessageDocument.jsx [hunk 6]
```diff
@@ -143,6 +143,23 @@ export const MessageDocument = connect(
                 </Button>
               )}
 
+              {messageDocumentHelper.showServePetitionButton && (
+                <Button
+                  link
+                  href={`/case-detail/${caseDetail.docketNumber}/petition-qc/${parentMessageId}`}
```
Comment (rachaelparris) [inline web-client/src/views/Messages/MessageDocument.jsx:149]: You have an href and an onClick here - probably just need the onClick I think?

File: web-client/src/presenter/sequences/navigateBackSequence.js (hunk 5 above)
Comment (matthopson) [reply web-client/src/presenter/sequences/navigateBackSequence.js:14]: Yup, already added to the story.

File: web-client/src/presenter/computeds/messageDocumentHelper.test.js (hunk 4 above)
Comment (matthopson) [reply web-client/src/presenter/computeds/messageDocumentHelper.test.js]: Yeah, I saw some more of those "is"'s in there and got them all.

File: web-client/src/router.js [hunk 7]
```diff
@@ -193,7 +193,7 @@ const router = {
     );
 
     registerRoute(
-      '/case-detail/*/petition-qc..',
+      '/case-detail/*/petition-qc',
```
Comment (rachaelparris) [inline web-client/src/router.js]: I'm a little concerned (maybe unnecessarily) about the removal of the `..` at the end of this route. Can you verify that the tab selection still works correctly? I think this was set up this way for the Edit buttons on the Review Petition screen to go to the petition QC screen with the correct tab selected.

Comment (matthopson) [reply web-client/src/router.js]: That's a good question. We need to be able to capture that parentMessageId with that other route and the `..` will thwart this. I'll dig around a little more and see if I can work out what happens when we move those.

Comment (matthopson) [reply web-client/src/router.js]: Flipping the routes around still works and allows for the tab to be set. Good catch!

File: web-client/src/presenter/computeds/messageDocumentHelper.test.js (hunk 4 above)
Comment (rachaelparris) [reply web-client/src/presenter/computeds/messageDocumentHelper.test.js]: Thank you!

**Review by rachaelparris — APPROVED** [review]

**Pull request merged**